### PR TITLE
Add particle filtering tutorial suggestions

### PR DESCRIPTION
**Pull request by hidmic**
### Proposed changes

Depends on #406. These are suggestions to the writing in #365. 

#### Type of change

- [ ] 🐛 Bugfix (change which fixes an issue)
- [ ] 🚀 Feature (change which adds functionality)
- [x] 📚 Documentation (change which fixes or extends documentation)

### Checklist

- [ ] Lint and unit tests (if any) pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] All commits have been signed for [DCO](https://developercertificate.org/)

### Additional comments

@Alondruck feel free to take the fragments you like and to drop the ones you don't.

### CODE REVIEW

**Review by Alondruck — COMMENTED** [review]
I am not very experienced in doing reviews, especially using GitHub, so if you don't understand any review I made, please let me know!

Overall, I see a significant improvement in the writing of the tutorial! 👏🏽

File: docs/tutorials/particle-filtering.md [hunk 1]
```diff
@@ -6,32 +6,27 @@
...
-where it constantly scanning the landmarks of the walls-doors world (the robot also moves only in a one-dimensional world). The last two graphs in the GIF showed above is a graphical representation of this world, where the blue bars represent the walls, the red bars represent the doors and the green bar is the robot.
+This tutorial takes after the one-dimensional particle filter example given in {cite}`thrun2005probabilistic{page 200, figure 8.11}`. There are walls and there are doors in our one-dimensional world, and our one-dimensional robot is free to move along this world. The last two plots in the animation shown above depict this world, where the blue bars are the walls, the red bars are the doors, and the green bar is the robot. While moving, our robot scans for doors and measures its distance to them. There is uncertainty in those measurements, however. There is also uncertainty in our robot proprioception (i.e. its sense of self-movement). We wish to estimate the robot position over time while acknowledging that uncertainty. So we will use a particle filter. The first two plots in the animation show how that position estimate or distribution evolves as the robot moves.
```
Comment (Alondruck) [inline docs/tutorials/particle-filtering.md:11]: My English and grammar are not the best, but check if this correction makes sense.
nit: 
```suggestion
This tutorial takes after the one-dimensional particle filter example given in {cite}`thrun2005probabilistic{page 200, figure 8.11}`. There are walls and there are doors in our one-dimensional world, and our one-dimensional robot is free to move along this world. The last two plots in the animation shown above depict this world, where the blue bars are the walls, the red bars are the doors, and the green bar is the robot. While moving, our robot scans for doors and measures its distance to them. There is uncertainty in those measurements, however. There is also uncertainty in our robot's proprioception (i.e. its sense of self-movement). We wish to estimate the robot's position over time while acknowledging that uncertainty. So we will use a particle filter. The first two plots in the animation show how that position estimate or distribution evolves as the robot moves.
```

File: docs/tutorials/particle-filtering.md [hunk 2]
```diff
@@ -6,32 +6,27 @@
...
 
-To implement this example in a code we will need to create and define different components such as the map, the motion and sensor models, the generation of the initial set of particles, the steps of the {abbr}`MCL (Monte Carlo Localization)` algorithm and the estimation of the robot's position.
+To implement this particle filter in Beluga, we will need a map, a motion model, a sensor model, and a prior or initial position estimate. We can then write an {abbr}`MCL (Monte Carlo Localization)` algorithm to estimate the robot position over time.
```
Comment (Alondruck) [inline docs/tutorials/particle-filtering.md:13]: same as before
nit:
```suggestion
To implement this particle filter in Beluga, we will need a map, a motion model, a sensor model, and a prior or initial position estimate. We can then write an {abbr}`MCL (Monte Carlo Localization)` algorithm to estimate the robot's position over time.
```

File: docs/tutorials/particle-filtering.md [hunk 3]
```diff
@@ -82,43 +78,49 @@ int main() {
...
     particles |= beluga::actions::reweight(std::execution::seq, sensor_model) | beluga::actions::normalize;
+    record.reweight = particles;
 
     particles |= beluga::views::sample |                                        //
                  ranges::views::take_exactly(parameters.number_of_particles) |  //
                  beluga::actions::assign;
+    record.resample = particles;
 
     const auto estimation = beluga::estimate(beluga::views::states(particles), beluga::views::weights(particles));
   }
```
Comment (Alondruck) [inline docs/tutorials/particle-filtering.md:126]: delete all the code related to records in this tutorial

```suggestion
    };

    particles |= beluga::actions::propagate(std::execution::seq, motion_model);

    particles |= beluga::actions::reweight(std::execution::seq, sensor_model) | beluga::actions::normalize;

    particles |= beluga::views::sample |                                        //
                 ranges::views::take_exactly(parameters.number_of_particles) |  //
                 beluga::actions::assign;

    const auto estimation = beluga::estimate(beluga::views::states(particles), beluga::views::weights(particles));
  }
```

File: docs/tutorials/particle-filtering.md [hunk 4]
```diff
@@ -185,16 +186,20 @@ struct Parameters {
...
-  std::vector<double> landmark_map{5, 12, 25, 37, 52, 55, 65, 74, 75, 87, 97};
+  /// Landmark positions in the simulated world.
+  std::vector<double> landmark_map;
+
+  /// Dataset path.
+  /**
+   * The dataset file is used to save the data produced by the simulation for posterior analisys.
+   */
+  std::filesystem::path record_path{"./record.yaml"};
 };
```
Comment (Alondruck) [inline docs/tutorials/particle-filtering.md:197]: Same as before
```suggestion
  /// Landmark positions in the simulated world.
  std::vector<double> landmark_map;
};
```

File: docs/tutorials/particle-filtering.md [hunk 5]
```diff
@@ -207,46 +212,48 @@ struct Parameters {
...
-* `ranges::views::transform(beluga::make_from_state<Particle>)` will transform each state into a weighted particle, as defined by `using Particle = std::tuple<double, beluga::Weight>;`.
-* `ranges::views::take_exactly(tutorial_params.number_of_particles)` defines the number of particles to take, using the `parameters.number_of_particles`.
-* `ranges::to<std::vector>` convert the range to a `std::vector<double>`.
+Note that:
+
+* we configure a `std::normal_distribution<double>` with `parameters.initial_position` and `parameters.initial_position_sigma` as mean and standard deviation, respectively.
+* `beluga::views::sample(initial_position_distribution)` samples robot states (ie. a scalar position) from the distribution.
+* `ranges::views::transform(beluga::make_from_state<Particle>)` transforms each state into a weighted particle, as defined by `using Particle = std::tuple<double, beluga::Weight>;`.
+* `ranges::views::take_exactly(tutorial_params.number_of_particles)` takes as many particles as `parameters.number_of_particles` specifies.
+* `ranges::to<std::vector>` turns the range into an `std::vector`.
```
Comment (Alondruck) [inline docs/tutorials/particle-filtering.md:239]: nit: just to maintain the same style as the following explanations.
```suggestion
Note that:

* we configure a `std::normal_distribution<double>` with `parameters.initial_position` and `parameters.initial_position_sigma` as mean and standard deviation, respectively.
* `beluga::views::sample` samples robot states (ie. a scalar position) from the distribution.
* `ranges::views::transform` transforms each state into a weighted particle, as defined by `using Particle = std::tuple<double, beluga::Weight>;`.
* `ranges::views::take_exactly` takes as many particles as `parameters.number_of_particles` specifies.
* `ranges::to<std::vector>` turns the range into an `std::vector`.
```

**Review by Alondruck — CHANGES_REQUESTED** [review]

File: docs/tutorials/particle-filtering.md (hunk 3 above)
Comment (hidmic) [reply docs/tutorials/particle-filtering.md:126]: Ahh, good catch. I thought I had removed it but clearly didn't.

**Comment by hidmic** [comment]
@Alondruck all suggestions accepted!

**Review by Alondruck — APPROVED** [review]
LGTM!

**Pull request merged**